opm/common/utility/FileSystem.hpp

```diff
@@ -0,0 +1,46 @@
+/*
+  Copyright 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef OPM_FILESYSTEM_HPP
+#define OPM_FILESYSTEM_HPP
+
+
+#if __cplusplus__ >= 201703L
+#include <filesystem>
+#else
+#include <experimental/filesystem>
+#endif
```

Comment:
On closer inspection, this particular logic doesn't actually work. There are two issues at play

1. We're referencing the wrong symbol.  The correct symbol name is `__cplusplus` with two leading underscores but no trailing underscores.

2. GCC 7 defines `__cplusplus` as `201703L` when using the `-std=c++17` flag:
    ```
    $ g++ --version
    g++ (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
    Copyright (C) 2017 Free Software Foundation, Inc.
    This is free software; see the source for copying conditions.  There is NO
    warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

    $ g++ -std=c++17 -x c++ -dM -E - </dev/null | grep __cplusplus
    #define __cplusplus 201703L
    ```

We'll need to write this check differently.  The following sort of achieves the goal, but it's particularly inelegant and I'd very much like a better way:
```diff
diff --git a/opm/common/utility/FileSystem.hpp b/opm/common/utility/FileSystem.hpp
index 603553e71..9e81f4d28 100644
--- a/opm/common/utility/FileSystem.hpp
+++ b/opm/common/utility/FileSystem.hpp
@@ -20,10 +20,11 @@
 #define OPM_FILESYSTEM_HPP
 
 
-#if __cplusplus__ >= 201703L
-#include <filesystem>
-#else
+#if (__cplusplus < 201703L) || \
+    (defined(__GNUC__) && (__GNUC__ < 8))
 #include <experimental/filesystem>
+#else
+#include <filesystem>
 #endif
 
 #include <string>
@@ -31,10 +32,11 @@
 
 namespace Opm
 {
-#if __cplusplus__ >= 201703L
-    namespace filesystem = std::filesystem;
-#else
+#if (__cplusplus < 201703L) || \
+    (defined(__GNUC__) && (__GNUC__ < 8))
     namespace filesystem = std::experimental::filesystem;
+#else
+    namespace filesystem = std::filesystem;
 #endif
```

Reply:
It might not be elegant, but if it is solid we should do it. After all, it shields Joe Developer (me!) well from these things.